### PR TITLE
Fix memory leaks

### PR DESCRIPTION
**Pull request by Slimakanzer**
This PR fixes syntax extension memory leaks.
In fact, this PR contains three independent parts (one commit for each part), so the reviewer has luxury to skip what he is not interested in:
- Fix memory leak when closing document. 
Basically, this is an unsubscribe from document events when it is actually closed.
- Fix memory leak when closing document window. 
Basically, this is an unsubscribe from document window events when it is closed.
- Fix unreasonable memory usage by the lexer/parser.
Previously, for each document, an object of the parser class was created, which stored a copy of the instruction set. Now lexer/parser are singletons.

### CODE REVIEW

File: VSRAD.Syntax/FunctionList/FunctionListProvider.cs [hunk 1]
```diff
@@ -116,26 +130,30 @@ private void ActiveDocumentChanged(IDocument activeDocument)
         private void ClearFunctionList()
         {
             _functionListControl.ClearList();
```
Comment (timlathy) [inline VSRAD.Syntax/FunctionList/FunctionListProvider.cs:132]: `_functionListControl` can be null at this point if the function list hasn't been open yet. When closing a document, we'll arrive here from `DocumentDisposed` and may hit a `NullReferenceException`. I suggest adding a null check.

File: VSRAD.Syntax/SyntaxHighlighter/ErrorHighlighter/SyntaxErrorHighlighterTaggerProvider.cs [hunk 2]
```diff
@@ -25,7 +25,7 @@ public SyntaxErrorHighlighterTaggerProvider(IDocumentFactory documentFactory)
             var document = _documentFactory.GetOrCreateDocument(buffer);
             if (document == null) return null;
 
-            return new SyntaxErrorHighlighterTagger(document.DocumentAnalysis) as ITagger<T>;
+            return new SyntaxErrorHighlighterTagger(document) as ITagger<T>;
```
Comment (timlathy) [inline VSRAD.Syntax/SyntaxHighlighter/ErrorHighlighter/SyntaxErrorHighlighterTaggerProvider.cs:28]: For some reason, `SyntaxErrorHighlighterTaggerProvider.CreateTagger` and `ErrorHighlighterTaggerProvider.CreateTagger` are called multiple times with the same `ITextBuffer` instance. Each time a document is open, a breakpoint placed here is hit three times. If you take a heap snapshot after opening several documents, you can see that there are more instances of `SyntaxErrorHighlighterTagger` and `ErrorHighlighterTagger` than there are of other taggers.

Replacing this line with the following solves the problem:
```
return buffer.Properties.GetOrCreateSingletonProperty(() => new SyntaxErrorHighlighterTagger(document) as ITagger<T>);
```

I couldn't pinpoint the exact conditions when `CreateTagger` is called (and the documentation does not mention this), so I think it's possible that multiple instances per buffer can be created for other taggers as well, the problem is just easier to reproduce with these two. I suggest you to use `GetOrCreateSingletonProperty` here and in other `CreateTagger` implementations.

File: VSRAD.Syntax/Core/LexerParserFactory.cs [hunk 3]
```diff
@@ -14,32 +15,50 @@ public struct LexerParser
...
     internal partial class DocumentFactory
     {
+        private Dictionary<AsmType, LexerParser> _asmSpecificLexerParser;
+
         private LexerParser? GetLexerParser(AsmType asmType)
         {
-            var instructionManager = _instructionManager.Value;
+            InitializeLexerParser();
+
+            if (_asmSpecificLexerParser.TryGetValue(asmType, out var lexerParser))
```
Comment (timlathy) [inline VSRAD.Syntax/Core/LexerParserFactory.cs:24]: (Subjective) I think that keeping the `switch` as it used to be and just replacing the `new` calls with `.Instance` accessors is a clearer approach that results in less code and less runtime overhead.

File: VSRAD.Syntax/FunctionList/FunctionListProvider.cs (hunk 1 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/FunctionList/FunctionListProvider.cs]: Thank you, fixed in 80f866a

File: VSRAD.Syntax/SyntaxHighlighter/ErrorHighlighter/SyntaxErrorHighlighterTaggerProvider.cs (hunk 2 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/SyntaxHighlighter/ErrorHighlighter/SyntaxErrorHighlighterTaggerProvider.cs]: Thank you! I assumed that the VS under the hood handles such a situation (for example peek definition creates new View for the same Buffer) and manages resources correctly, but as it turns out, it doesn't. 

Fixed in b08ab7f

File: VSRAD.Syntax/Core/LexerParserFactory.cs (hunk 3 above)
Comment (Slimakanzer) [reply VSRAD.Syntax/Core/LexerParserFactory.cs]: Thank you, I agree with you.
Fixed in 4dae221

**Pull request merged**